### PR TITLE
Recombine complete union of enum literals into original type (#9063)

### LINKED ISSUE #9063: [Feature request] Recombine complete union of enum literals into original type
This is feature that I think needs to be implemented before I can publish a PR for issue #6113.

```
from enum import Enum

class Color(Enum):
    RED = 1
    GREEN = 2
    BLUE = 3


x: Color
if x is Color.RED:
    pass
reveal_type(x)
```
*  Actual behavior/output
```
➜  mypy git:(master) ✗ python3 -m mypy leba_misc/test.py
leba_misc/test.py:12: note: Revealed type is 'Union[Literal[test.Color.RED], Literal[test.Color.GREEN], Literal[test.Color.BLUE]]'
```
* Desired behavior/output
```
leba_misc/test.py:12: note: Revealed type is 'test.Color'
```
While this isn't a bug per se, since the type is correct, but it's much less readable than displaying the original type. The problem becomes particularly bad with my implementation of #6113 , displaying `bool` as `Union[Literal[True], Literal[False]]`.

My thoughts on implementation are that it should be done either:
1. As an aesthetic change, only attempting to recombine literal unions for the purposes of user facing messages.
2. As part of the type checker.

Not sure what the best approach is, so I'd like to hear the mypy team's thoughts!

### PR DESCRIPTION
**Pull request by ethan-leba**
Closes #9063

### CODE REVIEW

**Comment by ethan-leba** [comment]
@JukkaL Is there any chance I could receive a review on this PR?

File: test-data/unit/check-enum.test [hunk 1]
```diff
@@ -632,6 +632,7 @@ elif x is Foo.C:
     reveal_type(x)  # N: Revealed type is 'Literal[__main__.Foo.C]'
 else:
     reveal_type(x)  # No output here: this branch is unreachable
+reveal_type(y) # N: Revealed type is '__main__.Foo'
```
Comment (TH3CHARLie) [inline test-data/unit/check-enum.test:635]: This doesn't make much sense to check the type of `y` here

File: test-data/unit/check-enum.test [hunk 2]
```diff
@@ -641,6 +642,7 @@ elif Foo.C is x:
...
```
Comment (TH3CHARLie) [inline test-data/unit/check-enum.test:645]: same as above

File: test-data/unit/check-enum.test [hunk 3]
```diff
@@ -877,7 +889,7 @@ def func(x: Union[int, None, Empty] = _empty) -> int:
                         # E: Unsupported left operand type for + ("Empty") \
                         # N: Left operand is of type "Union[int, None, Empty]"
     if x is _empty:
-        reveal_type(x)  # N: Revealed type is 'Literal[__main__.Empty.token]'
+        reveal_type(x)  # N: Revealed type is '__main__.Empty'
```
Comment (TH3CHARLie) [inline test-data/unit/check-enum.test:892]: I'm not sure whether should we display the original type when there's only one enum item. Displaying the only enum item seems like an acceptable result here.

**Review by TH3CHARLie — COMMENTED** [review]
Thanks for working on this! I like this cleaner message output. Left a few minor comments and/or uncertainties.

File: test-data/unit/check-enum.test (hunk 3 above)
Comment (ethan-leba) [reply test-data/unit/check-enum.test:892]: That's a fair point, I can add a special case in the logic for 1-item enums :+1:

Comment (TH3CHARLie) [reply test-data/unit/check-enum.test:892]: Cool!

**Review by TH3CHARLie — APPROVED** [review]
LGTM!

**Pull request merged**